packages/@lwc/template-compiler/src/codegen/static-element.ts

```diff
@@ -0,0 +1,217 @@
+    });
+
+    return staticNodes;
+}
+
+// The purpose of this function is to concatenate the contiguous text nodes into a single array
+// to simplify the traversing logic when generating static parts and serializing the element.
+export function transformStaticChildren(elm: StaticElement) {
+    const children = elm.children;
+    if (!children.length || !STATIC_ELEMENT_WITH_DYNAMIC_TEXT_SET.has(elm)) {
+        // The element either has no children or its children does not contain dynamic text.
+        return children;
+    }
+
+    if (STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE.has(elm)) {
+        return STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE.get(elm)!;
+    }
+
+    const result: (StaticChildNode | Text[])[] = [];
+    const len = children.length;
+
+    let current: StaticChildNode;
+    let next: StaticChildNode;
+    let contiguousTextNodes: Text[] | null = null;
+
+    for (let i = 0; i < len; i++) {
+        current = children[i];
+        if (!isText(current)) {
+            contiguousTextNodes = null;
+            result.push(current);
+        } else {
+            if (!isNull(contiguousTextNodes)) {
+                // Already in a contiguous text node chain
+                // All contiguous nodes represent an expression in the source, it's guaranteed by the parser.
+                contiguousTextNodes.push(current);
+            } else {
+                next = children[i + 1];
+                if (isExpression(current) || (!isUndefined(next) && isText(next))) {
+                    // Text nodes can appear as follows:
+                    // 1. A single text literal node.
+                    // 2. A single text expression node.
+                    // 3. Contiguous series of text nodes (literal/expression mixed) with at least 1 expression.
+                    // When there is an expression in the source, the text nodes are split into contiguous text nodes.
+                    // When there is no expression in the source, the text will appear as a single text literal.
+                    // We normalize all of the contiguous text nodes or single text expression to an array.
+                    // Single text literal nodes remain text nodes.
+                    // This is to normalize the traversal behavior when creating static parts and when serializing
+                    // the elements.
+                    contiguousTextNodes = [current];
+                }
+                // When contiguousTextNodes is null it is a single string literal.
+                result.push(contiguousTextNodes ?? current);
+            }
+        }
+    }
+
+    STATIC_ELEMENT_TO_DYNAMIC_TEXT_CHILDREN_CACHE.set(elm, result);
+
+    return result;
+}
+
+// Dynamic text is consolidated from individual text arrays into a single Text[].
+export const isDynamicText = (nodes: StaticChildNode | StaticChildNode[]): nodes is Text[] =>
+    isArray(nodes) && ArrayEvery.call(nodes, isText);
```

Comment:
This can technically be static text, correct?

Reply:
It can contain static text but if it contains an expression then it will always be in an array.

Static text = a single text literal node (not in an array).
Dynamic text = At least 1 text expression node + 0 or more text literal nodes (always in an array).

If it's just a static text node then it will not be in an array and will fail the array check in this test.

The reason we exclude static text is just to make it easier to distinguish between text nodes that need to be consolidated.

I'll add a comment to make it more clear.